File: src/main/java/每日一题/house_robber_II.java
package 每日一题;

public class house_robber_II {
    public static void main(String[] args) {
        int[] nums = new int[]{1, 2, 3, 1};
        System.out.println(rob(nums));
    }

    public static int rob(int[] nums) {
        int len = nums.length;
        if (len == 1) {
            return nums[0];
        }
        int[] dp1 = new int[len];
        int[] dp2 = new int[len];
        dp1[0] = nums[0];
        dp1[1] = Math.max(nums[0], nums[1]);
        dp2[0] = 0;
        dp2[1] = nums[1];
        //0 -> i-2
        for (int i = 2; i < len - 1; i++) {
            dp1[i] = Math.max(dp1[i - 1], dp1[i - 2] + nums[i]);
        }
        // 1 -> i-1
        for (int i = 2; i < len; i++) {
            dp2[i] = Math.max(dp2[i - 1], dp2[i - 2] + nums[i]);
        }
        return Math.max(dp1[len - 2], dp2[len - 1]);
    }
}

File: src/main/java/五毒/第六周/第一遍/house_robber_ii.go
package 第一遍

func robii(nums []int) int {
	length := len(nums)
	if length == 1 {
		return nums[0]
	}
	return max(maxRob(nums[:length-1]), maxRob(nums[1:]))
}

func maxRob(nums []int) int {
	length := len(nums)
	dp := make([][]int, length)
	var i int
	for i = 0; i < length; i++ {
		dp[i] = make([]int, 2)
	}
	dp[0][0] = 0
	dp[0][1] = nums[0]
	for i = 0; i < length; i++ {
		dp[i][0] = max(dp[i-1][0], dp[i-1][1])
		dp[i][1] = dp[i-1][0] + nums[i]
	}
	return max(dp[length-1][0], dp[length-1][1])
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}
